src/lib/common/ll.h

```diff
@@ -0,0 +1,119 @@
+#ifndef _LL_COMMON_H_
+#define _LL_COMMON_H_
+
+#include <stdint.h>
+#include <errno.h>
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+typedef struct LLNode_s
+{
+    struct LLNode_s *next;
+    struct LLNode_s *previous;
+    void *data;
+    uint32_t id;
+} LLNode_t;
+
+typedef struct
+{
+    LLNode_t *head;
+    LLNode_t *tail;
+    LLNode_t *p;
+} LL_t;
+
+typedef int (*LLTraverseCb_t)(void *data, void *arg);
+
+/******************************************************************************
+ * @brief Add A Node To A Linked List
+ *
+ * @details Please ensure that the node artifacts are not destroyed during the
+ *          use of the linked list. One way to ensure this is to make your
+ *          variables statically allocated.
+ *
+ * @param ll list to add node to
+ * @param node node to add to linked list
+ *
+ * @return 0 on success
+ * @return stderr on failure
+ *****************************************************************************/
+int LLNodeAdd(LL_t *ll, LLNode_t *node);
+
+/******************************************************************************
+ * @brief Get Data Element From Linked List
+ *
+ * @details Obtain data at specified id of linked list
+ *
+ * @param ll list to add node to
+ * @param id identifier of node to obtain data
+ * @param data data at specified identifier
+ *
+ * @return 0 on success
+ * @return stderr on failure
+ *****************************************************************************/
+int LLGetElement(LL_t *ll, uint32_t id, void **data);
+
+/******************************************************************************
+ * @brief Remove Node From Linked List
+ *
+ * @details Remove node at specified id in linked list
+ *
+ * @param ll list to delete node from
+ * @param id identifier of node to delete
+ *
+ * @return 0 on success
+ * @return stderr on failure
+ *****************************************************************************/
+int LLRemove(LL_t *ll, uint32_t id);
+
+/******************************************************************************
+ * @brief Pops Node Data On Linked List
+ *
+ * @details Utilizes an interator to pop off an element of a linked list and
+ *          retreive its data starting from the head. ll->p should be monitored
+ *          to determine if the iterator is at the end of the list
+ *          (ll->p == NULL).
+ *
+ * @param ll list to pop node from
+ * @param data data of popped node
+ *
+ * @return 0 on success
+ * @return stderr on failure
+ *****************************************************************************/
+int LLPop(LL_t *ll, void **data);
```

Comment:
Since "pop" as a name is typically used in stacks and removes the element last added to the end of the stack, I think this should have a different name. Since this function and `LLReset` are mainly provided to assist in traversal, my suggestion would be:

- Rename `LL_t` member `p` to `cursor`
- Rename `LLReset` to `LLResetCursor` ("reset" makes me think it will clear the list)
- Rename `LLPop` to `LLAdvanceCursor` or maybe `LLIterate` or `LLIterateCursor`
- Make both of those ☝️ functions private (remove from .h, make static in .c)

And thus make `LLTraverse` the only obvious way to iterate over the list, removing some potential confusion by future developers. Thoughts?

Reply:
I like that! Agreed on the naming scheme, this helps prevent confusion.
Making these functions static will also limit redundant API that developers do most likely will not utilize.

Reply:
Updated!